main.go

```diff
@@ -15,6 +15,138 @@
 
 package main
 
+import (
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/exec"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/spf13/cast"
+
+	"github.com/bank-vaults/secret-init/logger"
+	"github.com/bank-vaults/secret-init/providers"
+	"github.com/bank-vaults/secret-init/providers/vault"
+)
+
+var providersMap = map[string]providers.Provider{
+	// "aws": aws.NewAWSProvider(),
+	// "gcp": gcp.NewGCPProvider(),
+	"vault": vault.NewVaultProvider(),
+}
+
 func main() {
-	println("hello world")
+	logger := logger.SetupSlog()
+
+	var providerName string
+	if len(os.Args) >= 3 && os.Args[1] == "-p" {
+		providerName = os.Args[2]
+
+		// Remove the "-p {provider name}" argument from the command-line arguments
+		os.Args = append(os.Args[:1], os.Args[3:]...)
+	}
+
+	provider, found := providersMap[providerName]
+	if !found {
+		logger.Error("invalid provider specified.", slog.String("provider name", providerName))
+
+		os.Exit(1)
+	}
+
+	if len(os.Args) == 1 {
+		logger.Error("no command is given, vault-env can't determine the entrypoint (command), please specify it explicitly or let the webhook query it (see documentation)")
+
+		os.Exit(1)
+	}
+
+	daemonMode := cast.ToBool(os.Getenv("VAULT_ENV_DAEMON"))
+	delayExec := cast.ToDuration(os.Getenv("VAULT_ENV_DELAY"))
+	sigs := make(chan os.Signal, 1)
+
+	entrypointCmd := os.Args[1:]
+
+	binary, err := exec.LookPath(entrypointCmd[0])
+	if err != nil {
+		logger.Error("binary not found", slog.String("binary", entrypointCmd[0]))
+
+		os.Exit(1)
+	}
+
+	var envs []string
+	envs, err = provider.RetrieveSecrets(os.Environ())
+	if err != nil {
+		logger.Error("could not retrieve secrets from the provider.", err)
+
+		os.Exit(1)
+	}
+
+	if delayExec > 0 {
+		logger.Info(fmt.Sprintf("sleeping for %s...", delayExec))
+		time.Sleep(delayExec)
+	}
+
+	logger.Info("spawning process", slog.String("entrypoint", fmt.Sprint(entrypointCmd)))
+
+	if daemonMode {
```

Comment:
nit: Consider moving everything past this line to a separate function (for example `runEntrypointCmd`) with necessary arguments. This way, main could be used to extract argument, while the actual logic would be defined in this function.

Reply:
Let's wait for optimization until we understand what we are doing here. There are more than one issues with the current init code.